rest/data/mock_impl.go

```diff
@@ -0,0 +1,126 @@
+package data
+
+import (
+	"context"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/model/testresult"
+	restModel "github.com/evergreen-ci/evergreen/rest/model"
+	"github.com/evergreen-ci/utility"
+	"github.com/google/go-github/v34/github"
+	"github.com/pkg/errors"
+)
+
+type MockGitHubConnectorImpl struct {
+	Queue           map[string][]restModel.APICommitQueueItem
+	UserPermissions map[UserRepoInfo]string // map user to permission level in lieu of the Github API
+	CachedPatches   []restModel.APIPatch
+	Aliases         []restModel.APIProjectAlias
+	CachedTests     []testresult.TestResult
+	StoredError     error
+}
+
+func (pc *MockGitHubConnectorImpl) GetGitHubPR(ctx context.Context, owner, repo string, PRNum int) (*github.PullRequest, error) {
+	return &github.PullRequest{
+		User: &github.User{
+			ID:    github.Int64(1234),
+			Login: github.String("github.user"),
+		},
+		Base: &github.PullRequestBranch{
+			Ref: github.String("main"),
+		},
+		Head: &github.PullRequestBranch{
+			SHA: github.String("abcdef1234"),
+		},
+	}, nil
+}
+
+func (pc *MockGitHubConnectorImpl) AddPatchForPr(ctx context.Context, projectRef model.ProjectRef, prNum int, modules []restModel.APIModule, messageOverride string) (string, error) {
+	return "", nil
+}
+
+func (pc *MockGitHubConnectorImpl) IsAuthorizedToPatchAndMerge(ctx context.Context, settings *evergreen.Settings, args UserRepoInfo) (bool, error) {
+	_, err := settings.GetGithubOauthToken()
+	if err != nil {
+		return false, errors.Wrap(err, "can't get Github OAuth token from configuration")
+	}
+
+	permission, ok := pc.UserPermissions[args]
+	if !ok {
+		return false, nil
+	}
+	mergePermissions := []string{"admin", "write"}
+	hasPermission := utility.StringSliceContains(mergePermissions, permission)
+	return hasPermission, nil
+}
+
+func (pc *MockGitHubConnectorImpl) GetProjectFromFile(ctx context.Context, pRef model.ProjectRef, file string, token string) (model.ProjectInfo, error) {
+	config := `
+buildvariants:
+- name: v1
+  run_on: d
+  tasks:
+  - name: t1
+tasks:
+- name: t1
+`
+	p := &model.Project{}
+	opts := &model.GetProjectOpts{
+		Ref:          &pRef,
+		RemotePath:   file,
+		Token:        token,
+		ReadFileFrom: model.ReadFromLocal,
+	}
+	pp, pConfig, err := model.LoadProjectInto(ctx, []byte(config), opts, pRef.Id, p)
+	return model.ProjectInfo{
+		Project:             p,
+		IntermediateProject: pp,
+		Config:              pConfig,
+	}, err
+}
+
+func (mvc *MockGitHubConnectorImpl) CreateVersionFromConfig(ctx context.Context, projectInfo *model.ProjectInfo, metadata model.VersionMetadata, active bool) (*model.Version, error) {
+	return &model.Version{
+		Requester:         evergreen.GitTagRequester,
+		TriggeredByGitTag: metadata.GitTag,
+	}, nil
+}
+
+func (mtc *MockGitHubConnectorImpl) FindTestsByTaskId(opts FindTestsByTaskIdOpts) ([]testresult.TestResult, error) {
```

Comment:
Do tests need to be mocked? These things don't touch github.

Reply:
This doesn't need to be mocked per se but this test in particular has been quite difficult to mock and might need a large refactor to get it working (has to do with the fact that test ids are mgobson.ObjectIds and not strings, and this test runs into sorting issues when returning tests vs the hardcoded expected sequential list). I wrangled with it for a while and I think it's best to create a separate ticket, as you can see I reverted it to its mock behavior for now.